common/nodes/database-schema.xml

```diff
@@ -405,6 +405,50 @@ CREATE TABLE ib_memberships (
   INDEX (switch, part_name, interface)
 );
 
+<!-- Firmware -->
+<!-- need to drop in reverse order of creation because of FK constraints -->
+DROP TABLE IF EXISTS firmware;
+DROP TABLE IF EXISTS firmware_model;
+DROP TABLE IF EXISTS firmware_make;
+DROP TABLE IF EXISTS firmware_imp;
+
+CREATE TABLE firmware_imp (
+	id INT AUTO_INCREMENT PRIMARY KEY,
+	name VARCHAR(255) NOT NULL,
+	INDEX (name),
+	CONSTRAINT unique_name UNIQUE (name)
+);
+
+CREATE TABLE firmware_make (
+	id INT AUTO_INCREMENT PRIMARY KEY,
+	name VARCHAR(255) NOT NULL,
+	INDEX (name),
+	CONSTRAINT unique_name UNIQUE (name)
+);
+
+CREATE TABLE firmware_model (
+	id INT AUTO_INCREMENT PRIMARY KEY,
+	name VARCHAR(255) NOT NULL,
+	make_id INT NOT NULL,
+	FOREIGN KEY (make_id) REFERENCES firmware_make(id) ON DELETE RESTRICT,
+	INDEX (name),
+	CONSTRAINT unique_make_model UNIQUE (make_id, name)
+);
+
+CREATE TABLE firmware (
+	id INT AUTO_INCREMENT PRIMARY KEY,
+	model_id INT NOT NULL,
+	source VARCHAR(2048) NOT NULL,
+	version VARCHAR(255) NOT NULL,
+	hash_alg VARCHAR(255) NOT NULL default 'md5',
+	hash VARCHAR(2048) NOT NULL,
+	file VARCHAR(2048) NOT NULL,
+	imp_id INT DEFAULT NULL,
+	FOREIGN KEY (model_id) REFERENCES firmware_model(id) ON DELETE RESTRICT,
+	FOREIGN KEY (imp_id) REFERENCES firmware_imp(id) ON DELETE SET NULL,
+	INDEX (version),
+	CONSTRAINT unique_model_version UNIQUE (model_id, version)
```

Comment:
No need to name your constraints, you should be able to just change this line to:
```
UNIQUE (model_id, version)
```
If you plan on doing queries with both these columns in the WHERE clause, I would make it an unique index like so:
```
UNIQUE INDEX (model_id, version)
```

Reply:
I thought named constraints led to better error messages when the constraint was violated?

The things I've been reading said that a UNIQUE constraint makes a UNIQUE INDEX as well, but nothing appears to be terribly consistent on syntax so I'm not really sure what is real.

Reply:
`UNIQUE INDEX` is the same as `UNIQUE KEY`, but I believe you need either the `INDEX` or `KEY` part. But, it is easy enough to find out. Create the table with just the `UNIQUE` and then run `SHOW INDEX FROM my_cool_table;` inside of MySQL to see if an index is created.

Not sure about the constraint error messages, but we don't surface those anyways. But, I would think the error message would give you the columns failing the constraint, so the constraint name would just be extra noise.